python/tests/transfer_trigger_tests.py

```diff
@@ -210,3 +216,1349 @@ def test_trigger_manager_remove(self):
 
         for time, action in trigger_kinds_seprate():
             self.assertTrue(TM.trigger_map[Customer][time + action] == [])
+
+    # Do I need to test .extend??? 
+    ####################################################
+    """
+    Notes:
+        Read:
+            - Cannot read data of type x within a trigger attached to type x
+            - After read triggers dont execute unless the read is valid
+                - read isn't valid if you are reading an object that doesn't exist
+        Update:
+            - Before:
+                - inside the trigger: cannot change dimensions that are being changed in
+                  the actual update action that activated the trigger
+
+    Trigger test should be valid for Create, Read, Update, and Delete (CRUD) actions
+    Should check that:
+        Before:
+            1) Trigger arguments: old, new, current - are the correct values
+            2) Using BlockAction prevents the CRUD action from taking place (only when before)
+            3) Before triggers can alter demensions of an object that is being CRUD
+            4) Any trigger can access/change [similar] objects in dataframe
+            5) Any trigger can access/change [different] objects in dataframe
+            6) Any trigger can insert objects into the dataframe
+            7) That overlapping triggers execute in the right order
+        After
+            1) Trigger arguments: old, new, current - are the correct values
+            2) Using BlockAction doesn't do anything to the object of concern
+            3) After triggers cannot alter demensions of an object that is being CRUD
+            4) Any trigger can access/change [similar] objects in dataframe
+            5) Any trigger can access/change [different] objects in dataframe
+            6) Any trigger can insert objects into the dataframe
+            7) That overlapping triggers execute in the right order
+   """
+
+    #######################
+    #### before create ####
+
+    def test_bc_1(self):
+        """
+        Tests that before create trigger arguments are the correct values
+        dataframe == dataframe object
+        new == object trigger is dealing with
+        old == None
+        current == None
+        """
+        @trigger(Transaction, TriggerTime.before, TriggerAction.create)
+        def before_create(dataframe, new, old, current):
+            # Need to use this type of assert because this is executes outside this class
+            assert (str(type(dataframe)) ==
```

Comment:
Create a datastructure outside the trigger but inside `test_bc_1`  to hold the data that the before_create trigger is going to perform.

Then after the trigger is executed, check this datastructure with normal asserts to see if all the things that the trigger looks at and produces are in the datastructure. That way we dont break test flow by using regular assert, and we can still check what we need to check.